File: filltrimino.py
#!/usr/bin/python

# Program to fill 2^n by 2^n square with a hole in it using triminos   |A|
#                                                                    |A|B|

import random

def filltrimino(board, size, hole, start):
    (x, y) = hole
    (p, q) = start
    if x < p+size/2 and y < q+size/2:
        board[p+size/2][q+size/2] = 'B'
        board[p+size/2-1][q+size/2] = 'A'
        board[p+size/2][q+size/2-1] = 'A'
        if size > 2:
            filltrimino(board, size/2, hole, start)
            filltrimino(board, size/2, (p+size/2, q+size/2), (p+size/2, q+size/2))
            filltrimino(board, size/2, (p+size/2-1, q+size/2), (p, q+size/2))
            filltrimino(board, size/2, (p+size/2, q+size/2-1), (p+size/2, q))
    elif x >= p+size/2 and y < q+size/2:
        board[p+size/2-1][q+size/2-1] = 'A'
        board[p+size/2-1][q+size/2] = 'B'
        board[p+size/2][q+size/2] = 'A'
        if size > 2:
            filltrimino(board, size/2, hole, (p+size/2, q))
            filltrimino(board, size/2, (p+size/2-1, q+size/2-1), start)
            filltrimino(board, size/2, (p+size/2-1, q+size/2), (p, q+size/2))
            filltrimino(board, size/2, (p+size/2, q+size/2), (p+size/2, q+size/2))
    elif x < p+size/2 and y >= q+size/2:
        board[p+size/2-1][q+size/2-1] = 'A'
        board[p+size/2][q+size/2-1] = 'B'
        board[p+size/2][q+size/2] = 'A'
        if size > 2:
            filltrimino(board, size/2, hole, (p, q+size/2))
            filltrimino(board, size/2, (p+size/2, q+size/2), (p+size/2, q+size/2))
            filltrimino(board, size/2, (p+size/2-1, q+size/2-1), start)
            filltrimino(board, size/2, (p+size/2, q+size/2-1), (p+size/2, q))
    elif x >= p+size/2 and y >= q+size/2:
        board[p+size/2-1][q+size/2-1] = 'B'
        board[p+size/2-1][q+size/2] = 'A'
        board[p+size/2][q+size/2-1] = 'A'
        if size > 2:
            filltrimino(board, size/2, hole, (p+size/2, q+size/2))
            filltrimino(board, size/2, (p+size/2-1, q+size/2-1), start)
            filltrimino(board, size/2, (p+size/2-1, q+size/2), (p, q+size/2))
            filltrimino(board, size/2, (p+size/2, q+size/2-1), (p+size/2, q))

n = random.randint(1,3)
size = 2**n
x = random.randint(1, size)
y = random.randint(1, size)

print "Filling triminos for square", size, "by", size, "with a hole at (", x, ",", y, ")"

board = []
for i in range(size+1):
    row = []
    for j in range(size+1):
        row.append('X')
    board.append(row)

filltrimino(board, size, (x, y), (1,1));

for i in range(1, size+1):
    print str(board[i][1:])[1:-1]

File: filltrimino.js
function getRandomColor()
{
 var color =  '#'+Math.floor(Math.random()*0xffffff).toString(16);
 return color;
}

var count = 0;
function filltrimino(board, size, hole, start)
{
  var x = hole[0]; var y = hole[1];
  var p = start[0]; var q = start[1];
  count = count + 1;
  var color = getRandomColor();

  if (x < p+size/2 && y < q+size/2) {
        board[p+size/2][q+size/2] = [color, count];
        board[p+size/2-1][q+size/2] = [color, count];
        board[p+size/2][q+size/2-1] = [color, count];
        if (size > 2) {
            filltrimino(board, size/2, hole, start);
            filltrimino(board, size/2, [p+size/2, q+size/2], [p+size/2, q+size/2]);
            filltrimino(board, size/2, [p+size/2-1, q+size/2], [p, q+size/2]);
            filltrimino(board, size/2, [p+size/2, q+size/2-1], [p+size/2, q]);
        }
    }
    else if (x >= p+size/2 && y < q+size/2) {
        board[p+size/2-1][q+size/2-1] = [color, count];
        board[p+size/2-1][q+size/2] = [color, count];
        board[p+size/2][q+size/2] = [color, count];
        if (size > 2) {
            filltrimino(board, size/2, hole, [p+size/2, q]);
            filltrimino(board, size/2, [p+size/2-1, q+size/2-1], start);
            filltrimino(board, size/2, [p+size/2-1, q+size/2], [p, q+size/2]);
            filltrimino(board, size/2, [p+size/2, q+size/2], [p+size/2, q+size/2]);
        }
    }
    else if (x < p+size/2 && y >= q+size/2) {
        board[p+size/2-1][q+size/2-1] = [color, count];
        board[p+size/2][q+size/2-1] = [color, count];
        board[p+size/2][q+size/2] = [color, count];
        if (size > 2) {
            filltrimino(board, size/2, hole, [p, q+size/2]);
            filltrimino(board, size/2, [p+size/2, q+size/2], [p+size/2, q+size/2]);
            filltrimino(board, size/2, [p+size/2-1, q+size/2-1], start);
            filltrimino(board, size/2, [p+size/2, q+size/2-1], [p+size/2, q]);
        }
    }
    else if (x >= p+size/2 && y >= q+size/2) {
        board[p+size/2-1][q+size/2-1] = [color, count];
        board[p+size/2-1][q+size/2] = [color, count];
        board[p+size/2][q+size/2-1] = [color, count];
        if (size > 2) {
            filltrimino(board, size/2, hole, [p+size/2, q+size/2]);
            filltrimino(board, size/2, [p+size/2-1, q+size/2-1], start);
            filltrimino(board, size/2, [p+size/2-1, q+size/2], [p, q+size/2]);
            filltrimino(board, size/2, [p+size/2, q+size/2-1], [p+size/2, q]);
        }
    }
    return;
}

var n = Math.floor((Math.random() * 4) + 1);
var size = Math.pow(2, n);
var x = Math.floor((Math.random() * size) + 1);
var y = Math.floor((Math.random() * size) + 1);

var board = new Array(size+1);

for (var i = 0; i < size+1; i++) {
  board[i] = new Array(size+1);
  for (var j = 0; j < size+1; j++) {
     board[i][j] = ["white", 0];
  }
}

board[x][y] = ["snow", 'X'];

filltrimino(board, size, [x,y], [1,1]);

var chessBoard = document.getElementById("chessBoard");

for (var i=1; i<size+1; i++){
    var row = chessBoard.appendChild(document.createElement("div"));
    for (var j=1; j<size+1; j++){
    		var t = document.createElement("div");
        var numtxt = document.createTextNode(board[i][j][1]);
        t.style.backgroundColor = board[i][j][0];
        t.appendChild(numtxt);
        row.appendChild(t);
    }
}
